rali/rali/source/coco_file_source_reader.cpp

```diff
@@ -0,0 +1,252 @@
+{
+    if (_loop)
+        return _file_names.size();
+
+    int ret = ((int)_file_names.size() - _read_counter);
+    return ((ret < 0) ? 0 : ret);
+}
+
+Reader::Status COCOFileSourceReader::initialize(ReaderConfig desc)
+{
+    auto ret = Reader::Status::OK;
+    _file_id = 0;
+    _folder_path = desc.path();
+    _json_path = desc.json_path();
+    _shard_id = desc.get_shard_id();
+    _shard_count = desc.get_shard_count();
+    _batch_count = desc.get_batch_size();
+    _loop = desc.loop();
+    _shuffle = desc.shuffle();
+
+    if(_json_path == "")
+    {
+        std::cout<<"\n _json_path has to be set manually";
+        exit(0);
+    }
+    MetaDataConfig config(MetaDataType::BoundingBox, MetaDataReaderType::COCO_META_DATA_READER, _json_path);
+    _meta_data_graph = create_meta_data_graph(config);
+    _meta_data_reader = create_meta_data_reader(config);
+    _meta_data_reader->init(config);
+    _meta_data_reader->read_all(_json_path);
+    ret = subfolder_reading();
+    // the following code is required to make every shard the same size:: required for multi-gpu training
+    if (_shard_count > 1 && _batch_count > 1) {
+        int _num_batches = _file_names.size()/_batch_count;
+        int max_batches_per_shard = (_file_count_all_shards + _shard_count-1)/_shard_count;
+        max_batches_per_shard = (max_batches_per_shard + _batch_count-1)/_batch_count;
+        if (_num_batches < max_batches_per_shard) {
+            replicate_last_batch_to_pad_partial_shard();
+        }
+    }
+    //shuffle dataset if set
+    _shuffle_time.start();
+    if (ret == Reader::Status::OK && _shuffle)
+        std::random_shuffle(_file_names.begin(), _file_names.end());
+    _shuffle_time.end();
+    return ret;
+}
+
+void COCOFileSourceReader::incremenet_read_ptr()
+{
+    _read_counter++;
+    _curr_file_idx = (_curr_file_idx + 1) % _file_names.size();
+}
+size_t COCOFileSourceReader::open()
+{
+    auto file_path = _file_names[_curr_file_idx]; // Get next file name
+    incremenet_read_ptr();
+    _last_id = file_path;
+    auto last_slash_idx = _last_id.find_last_of("\\/");
+    if (std::string::npos != last_slash_idx)
+    {
+        _last_id.erase(0, last_slash_idx + 1);
+    }
+
+    _current_fPtr = fopen(file_path.c_str(), "rb"); // Open the file,
+
+    if (!_current_fPtr) // Check if it is ready for reading
+        return 0;
+
+    fseek(_current_fPtr, 0, SEEK_END); // Take the file read pointer to the end
+
+    _current_file_size = ftell(_current_fPtr); // Check how many bytes are there between and the current read pointer position (end of the file)
+
+    if (_current_file_size == 0)
+    { // If file is empty continue
+        fclose(_current_fPtr);
+        _current_fPtr = nullptr;
+        return 0;
+    }
+
+    fseek(_current_fPtr, 0, SEEK_SET); // Take the file pointer back to the start
+
+    return _current_file_size;
+}
+
+size_t COCOFileSourceReader::read(unsigned char *buf, size_t read_size)
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5976200)